src/hybrid_reader.c

```diff
@@ -0,0 +1,299 @@
+    }
+    // Otherwise, advance the iterator pointing to the lower id.
+    else if (cur_vec_res->docId > cur_child_res->docId && IITER_HAS_NEXT(hr->child)) {
+      int rc = hr->child->SkipTo(hr->child->ctx, cur_vec_res->docId, &cur_child_res);
+      if (rc == INDEXREAD_NOTFOUND) {
+        hr->child->Read(hr->child->ctx, &cur_child_res);
+      }
+    } else if (VecSimQueryResult_IteratorHasNext(vecsim_iter)){
+      HR_SkipToInBatch(hr, cur_child_res->docId, &cur_vec_res);
+    } else {
+      break; // both iterators are depleted.
+    }
+  }
+  IndexResult_Free(cur_vec_res);
+}
+
+void prepareResults(HybridIterator *hr) {
+    if (hr->mode == STANDARD_KNN) {
+      hr->list =
+          VecSimIndex_TopKQuery(hr->index, hr->query.vector, hr->query.k, &(hr->runtimeParams), hr->query.order);
+      hr->iter = VecSimQueryResult_List_GetIterator(hr->list);
+    return;
+  }
+
+  if (hr->mode == HYBRID_ADHOC_BF) {
+    // Go over child_it results, compute distances, sort and store results in topResults.
+    return;
+  }
+  // Batches mode.
+  VecSimBatchIterator *batch_it = VecSimBatchIterator_New(hr->index, hr->query.vector);
+  float upper_bound = INFINITY;
+  while (VecSimBatchIterator_HasNext(batch_it)) {
+    size_t batch_size = hr->query.k;  // todo: add heuristics here
+    if (hr->list) {
+      VecSimQueryResult_Free(hr->list);
+    }
+    if (hr->iter) {
+      VecSimQueryResult_IteratorFree(hr->iter);
+    }
+    // Get the next batch.
+    hr->list = VecSimBatchIterator_Next(batch_it, batch_size, BY_ID);
+    hr->iter = VecSimQueryResult_List_GetIterator(hr->list);
+    hr->child->Rewind(hr->child->ctx);
+
+    // Go over both iterators and save mutual results in the heap.
+    alternatingIterate(hr, hr->iter, &upper_bound);
+    if (heap_count(hr->topResults) == hr->query.k) {
+      break;
+    }
+  }
+  VecSimBatchIterator_Free(batch_it);
+}
+
+static int HR_HasNext(void *ctx) {
+  HybridIterator *hr = ctx;
+  return hr->base.isValid;
+}
+
+static int HR_ReadUnsorted(void *ctx, RSIndexResult **hit) {
+  HybridIterator *hr = ctx;
+  if (!hr->resultsPrepared) {
+    prepareResults(hr);
+    hr->resultsPrepared = true;
+  }
+  if (!HR_HasNext(ctx)) {
+    return INDEXREAD_EOF;
+  }
+  if (hr->mode == STANDARD_KNN) {
+    *hit = hr->base.current;
+    if (HR_ReadInBatch(hr, hit) == INDEXREAD_EOF) {
+      goto eof;
+    }
+  } else {
+    if (heap_count(hr->topResults) == 0) {
+      goto eof;
+    }
+    *hit = heap_poll(hr->topResults);
+    hr->returnedResults = array_append(hr->returnedResults, *hit);
+  }
+  hr->lastDocId = (*hit)->docId;
+  return INDEXREAD_OK;
+
+eof:
+  hr->base.isValid = false;
+  return INDEXREAD_EOF;
+}
+
+static bool UseBF(size_t T, TopKVectorQuery query, VecSimIndex *index) {
+  // todo: have more sophisticated heuristics here, currently this is disabled.
+  // return (float)T < (0.05 * (float)VecSimIndex_IndexSize(index));
+  return false;
+}
+
+static size_t HR_NumEstimated(void *ctx) {
```

Comment:
do we test this?

Reply:
no. I will add unit tests for the iterator...